Document policy for adding ecosystem/integration test jobs/components to rust-lang/rust CI

jieyouxu opened the pull request:
MCP: https://github.com/rust-lang/compiler-team/issues/845
MCP thread: [#t-compiler/major changes > Policy: Require MCP and relevant team nom… compiler-team#845](https://rust-lang.zulipchat.com/#narrow/channel/233931-t-compiler.2Fmajor-changes/topic/Policy.3A.20Require.20MCP.20and.20relevant.20team.20nom.E2.80.A6.20compiler-team.23845/with/506935487)
rust-lang/rust issue: https://github.com/rust-lang/rust/issues/137960

Following discussions in https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/RFL.20CI.20Job.

r? @davidtwco (or @wesleywiser)

[Rendered](https://github.com/rust-lang/rust-forge/blob/master/src/SUMMARY.md)

jieyouxu commented:
Blocked on opening an MCP for this first.
@rustbot blocked

jieyouxu commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
```diff
@@ -0,0 +1,88 @@
+- If the custom test job/component will block on failure, why does it need to block?
+- If the custom test job/component will not block on failure initially but is intended to eventually become blocking, why so? And when will it become blocking?
+
+### Test job/component maintainers
+
+The proposed custom test job/component for [rust-lang/rust] CI must have at least one dedicated maintainer. The test job/component maintainers understand that they will be pinged or otherwise contacted about the custom test job/component, particularly for (but not limited to) its failures. Please list who will be maintaining this test job/component here.
+
+**We, the custom test job/component maintainers, acknowledge and consent that we will be pinged about the custom test job/component failures or be consulted about it in general.**
```
Discussion: I haven't carried over the test job/component ping group being **mandatory** over to this draft yet. I can modify the draft to require a ping group if people find that is necessary.

davidtwco replied:
I think a mandatory ping group is good.

jieyouxu replied:
It turns out that only team members and certain WG-* members can even use `@rustbot ping xxx`, which IMO makes mandatory ping groups mostly useless. Non-team/wg-members can't ping `@rust-lang/xxx` nor `@rustbot ping xxx` anyway, while org members can directly `@rust-lang/xxx` or just `@` the test job maintainers directly.

jieyouxu replied:
I removed mentions of ping groups in (round 2).

jieyouxu commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
```diff
@@ -0,0 +1,88 @@
+- Why does this custom test job/component need to be part of the [rust-lang/rust] PR and/or Full Merge CI?
+- If the custom test job/component will block on failure, why does it need to block?
+- If the custom test job/component will not block on failure initially but is intended to eventually become blocking, why so? And when will it become blocking?
+
```
TODO: missing a section for

> We would also like that process to contain a step where it must outlined what such a project is depending on so we can better understand what we're getting into.
Are they using unstable features? Do they knowingly depend on implementation details? Do they plan to expand or shrink such dependencies at a later point?

and

> RfL can provide valuable feedback, but there's also a big risk if a feature isnt actually ready for such exposure yet or must be completely reworked -- that'll be a bunch of work to migrate the kernel and also a bunch of reputational damage

(See [#t-compiler/major changes > Policy: Require MCP and relevant team nom… compiler-team#845 @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/233931-t-compiler.2Fmajor-changes/topic/Policy.3A.20Require.20MCP.20and.20relevant.20team.20nom.E2.80.A6.20compiler-team.23845/near/503592508))

jieyouxu replied:
Added some preliminary wording to address these questions.

Kobzol commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
```diff
@@ -0,0 +1,96 @@
+For example, this includes but is not limited to:
+
+- Ecosystem test jobs, such as Rust for Linux or Fuchsia;
+- Custom codegen backend test components, such as `codegen_cranelift` or `codegen_gcc`, which are not part of the "main" distributed toolchain.
```
`cg_clif` has been shipping with rustup for quite some time, so I think that's a bit different from `cg_gcc`. If we ship the thing with rustup, I think that it essentially becomes the "main" distributed toolchain.

jieyouxu replied:
Makes sense to me. I think I will remove cg_clif from this list since it's indeed part of the "main" distributed toolchain. (We're still setting up a ping group for cg_clif in https://github.com/rust-lang/rust/pull/138120, but that's mostly for consultation purposes.)

Kobzol replied:
Yeah, these are two separate things (is main toolchain vs has more than ~one person who can fix it :) ).

jieyouxu replied:
Remove cg_clif from this list.

Kobzol commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
```diff
@@ -0,0 +1,96 @@
+
+## Process for adding a custom test job to rust-lang/rust CI
+
+![Flowchart of the process for adding a custom test job/component to rust-lang/rust CI](./custom-test-jobs/test-job-flow.png)
```
I don't suppose that we could build this e.g. with Mermaid or at least SVG? Having this be a PNG (without some source) essentially means that it won't be ever possible to be changed without redoing it from scratch :)

jieyouxu replied:
It can be, this is mostly *a* diagram. Happy to change it to whatever format that's more maintainble, I just used a naive draw.io flowchart for illustrative purposes.

jieyouxu replied:
The exported SVG form is

![test-job-flow drawio](https://github.com/user-attachments/assets/2a6634e3-dd06-4973-b5bd-900d665026b7)

but yeah.

Kobzol replied:
We could just commit the SVG, I think. Could be rendered a bit differently on different browsers, but that shouldn't be such an issue here.

jieyouxu replied:
Yeah, that seems at least easier to edit than PNG, although I can't say I'm an expert at editing an SVG either.

jieyouxu replied:
Committed the SVG instead of the PNG.

Kobzol reviewed (commented):
Looks great, thanks. The diagram kind of shows how desynchronized are the decision processes between t-compiler and t-libs (and other teams) currently :) Left some comments.

jieyouxu commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
```diff
@@ -0,0 +1,96 @@
+
+### Test job/component maintainers
+
+The proposed custom test job/component for [rust-lang/rust] CI must have at least one dedicated maintainer. The test job/component maintainers understand that they will be pinged or otherwise contacted about the custom test job/component, particularly for (but not limited to) its failures. Please list who will be maintaining this test job/component here.
```
Discussion: we could also be very upfront about what happens if a custom test job/component no longer having a dedicated maintainer, that:

- We may ask around if anyone wants to sign up as a dedicated maintainer for this specific test job/component
- If we can't find a dedicated maintainer, and that e.g. compiler/libs/infra (or if the test job can break tools in r-l/r, e.g. cargo/rustdoc/...) find the test job/component to not pull its weight / cause significant burden / is no longer relevant, then the test job/component may be removed.

davidtwco replied:
I think that would be a reasonable addition.

jieyouxu replied:
Added this remark about unmaintained job/components may be removed in (round 2).

jieyouxu commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
```diff
@@ -0,0 +1,96 @@
+
+can introduce a lot of friction and frustration for other contributors who now will have to deal with the failure or understand why the test job failed in the first place, which may now be blocking their PR. Everyone utilizing the [rust-lang/rust] CI is expected to use it responsibly.
+
+To help with this, please follow the process described below if you'd like to propose adding a custom test job/component to the [rust-lang/rust] CI, and copy and fill out the Custom Test Job/Component Policy.
```
Problem (for myself): I need to explicitly note here (as described in the MCP) that this policy is intended to be **retroactive**.

davidtwco replied:
I'm not sure that the policy needs to say it is retrospective, as that has no bearing on anyone reading this later to propose another job. We can just implement this policy for previously added jobs.

jieyouxu replied:
That is true, that only really makes sense for the MCP.

jieyouxu commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
```diff
@@ -0,0 +1,96 @@
+
+## Scope
+
+This policy is applicable to proposals for adding new test jobs or adjusting existing test jobs to build additional custom things that are *not* shipped to the end users, but may still cause [rust-lang/rust] PR CI or Full CI (the "CI") to fail, or produce a failure message, which may impact compiler/library/tools contributors who also use the [rust-lang/rust] CI.
+
+For example, this includes but is not limited to:
+
+- Ecosystem test jobs, such as Rust for Linux or Fuchsia;
+- Custom codegen backend test components, such as `codegen_gcc`, which are not part of the "main" distributed toolchain.
```
Discussion: I don't really have a good short phrase to describe "custom test job/component", please do lmw if anyone has ideas.

davidtwco replied:
maybe "ecosystem and integration testing"?

jieyouxu replied:
Changed terminology to stick with "ecosystem/integration test jobs/components" in (round 2). Filename was not renamed for review purposes.

jieyouxu commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
```diff
@@ -0,0 +1,96 @@
+
+For example, this includes but is not limited to:
+
+- Ecosystem test jobs, such as Rust for Linux or Fuchsia;
+- Custom codegen backend test components, such as `codegen_gcc`, which are not part of the "main" distributed toolchain.
```
Remark: this scope is preliminary (and reflects current test job/component failures being difficult to handle that motivated this policy), but can be amended in the future to adapt to future needs.

jieyouxu replied:
I slightly reworded this to just be "ecosystem/integration test job/component".

jieyouxu commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
```diff
@@ -0,0 +1,96 @@
+
+## Process for adding a custom test job to rust-lang/rust CI
+
+![Flowchart of the process for adding a custom test job/component to rust-lang/rust CI](./custom-test-jobs/test-job-flow.svg)
```
Problem (for myself): I need to transcribe the diagram as text form; the digram is only illustrative and cannot be used as a substitute for proper prose.

jieyouxu replied:
Transcribed and fixed diagram, including an early step to ask infra if there's capacity (round 2).

jieyouxu commented on src/compiler/proposals-and-stabilization/custom-test-jobs/test-job-flow.svg:
TODO: adjust flow to "ask infra ahead of time in zulip stream if there would be sufficient capacity"

jieyouxu replied:
Fixed in (round 2).

davidtwco commented on src/compiler/proposals-and-stabilization.md:
```diff
@@ -329,9 +329,15 @@ circumstance.
 
+### Adding new custom test jobs/components to rust-lang/rust CI
+
+See [Adding custom test jobs or adding custom test components to rust-lang/rust CI](./proposals-and-stabilization/custom-test-jobs.md).
```
```suggestion
See [*Adding custom test jobs or adding custom test components to rust-lang/rust CI*](./proposals-and-stabilization/custom-test-jobs.md).
```

jieyouxu replied:
Fixed in (round 2).

davidtwco commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
```diff
@@ -0,0 +1,96 @@
+# Adding custom test jobs or adding custom test components to rust-lang/rust CI
```
nit: there's line-wrapping on the other files in our section, could you hard wrap this file too?

jieyouxu replied:
Hard-wrapped to 100 columns in an intermediate commit before (round 2).

davidtwco commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
```diff
@@ -0,0 +1,96 @@
+
+can introduce a lot of friction and frustration for other contributors who now will have to deal with the failure or understand why the test job failed in the first place, which may now be blocking their PR. Everyone utilizing the [rust-lang/rust] CI is expected to use it responsibly.
+
+To help with this, please follow the process described below if you'd like to propose adding a custom test job/component to the [rust-lang/rust] CI, and copy and fill out the Custom Test Job/Component Policy.
```
```suggestion
To help with this, please follow the process described below if you'd like to propose adding a custom test job/component to the [rust-lang/rust] CI, and copy and fill out the *Custom Test Job/Component Policy*.
```

davidtwco reviewed (approved):
Some nits, but this looks great.

jieyouxu commented:
Changes since last review:

- Hard-wrapped to 100 columns.
- Changed terminology to use "ecosystem/integration test jobs/components" instead.
- Spelled out the process (instead of just having the diagram).
- Adjusted process (text and diagram) to ask infra about capacity first (discussed in [#t-infra > Policy for adding custom test jobs/components to r-l/r CI](https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/Policy.20for.20adding.20custom.20test.20jobs.2Fcomponents.20to.20r-l.2Fr.20CI/with/507219752)).
- Dropped mentions about ping groups because only team members + certain WG members can use triagebot ping group anyway.
- Adjusted scope wording.

jieyouxu commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
FIXME(jieyouxu): rename this file to sth like `ecosystem-integration-tests.md` and fix links

davidtwco replied:
This can be merged once you do this.

jieyouxu replied:
Fixed filename and links.

jieyouxu commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
```diff
@@ -0,0 +1,157 @@
+1. The Infrastructure Team will review the implementation PR. Once approved, the
+   ecosystem/integration test job/component will run in [rust-lang/rust] PR CI or Full Merge CI.
+
+## Ecosystem and Integration Test Job/Component Policy
+
+*Please copy and fill out this template as part of the MCP.*
```
FIXME(jieyouxu): make it clearer what is part of the template vs what the person is expected to replace and fill out

jieyouxu replied:
Fixed by a using quoteblocks for policy content itself, italicized placeholder text for things the MCP author is expected to include.

jieyouxu commented:
Changes since last push:

- I significantly reformatted the policy template to make it much more edit-friendly, as I see in https://github.com/rust-lang/compiler-team/issues/855 that it was hard to work with.

Kobzol commented on src/compiler/proposals-and-stabilization/custom-test-jobs.md:
```diff
@@ -0,0 +1,248 @@
+1. Ask the Infrastructure Team on zulip
+  ([#t-infra](https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra)) to check if there
+  would be capacity for the proposed test job/component.
+1. Propose using a [Major Change Proposal
```
I know that Markdown might render it correctly, but shouldn't ths be `1), 2), 3)` etc.? :)

jieyouxu replied:
Yeah, it will render correctly, but I can also change it to 1) 2) 3) in source :D

jieyouxu replied:
Fixed it to use 1. 2. 3. instead.

jieyouxu commented:
I'll fix the numbering and the file link renames

jieyouxu commented:
File is renamed with links fixed
